Replace custom conditions article with Validate your configuration (#36793)
hashicorp/terraform: website/docs/language/mitigate-risk/validate.mdx

The following example uses a precondition to verify that the AMI configured for an`aws_instance` uses the `x86_64` CPU architecture.

```hcl
resource "aws_instance" "example" {
  instance_type = "t3.micro"
  ami           = data.aws_ami.example.id

  lifecycle {
    # The AMI ID must refer to an AMI that contains an operating system
    # for the `x86_64` architecture.
    precondition {
      condition     = data.aws_ami.example.architecture == "x86_64"
      error_message = "The selected AMI must be for the x86_64 architecture."
    }
  }
}
```

The precondition detects if the caller accidentally selected an AMI for a different architecture, which may not be able to run the software this instance hosts. Terraform evaluates the precondition while it builds its plan, and if the precondition fails Terraform throws an error with the `error_message` argument and stops the current operation. Refer to the [resource configuration reference](/terraform/language/block/resource) for more examples of using a precondition block.

> An `output` block can also include a precondition to verify a module's output. You can use preconditions on outputs to prevent Terraform from saving an invalid value in your state, or to preserve a valid value from the previous apply.

Author: @ritsok Following up on our discussion:

You said: "Not sure if I misunderstand, but this seems like an antipattern - do we have an example of this somewhere?"

And in the current article, we say "You can use preconditions to prevent Terraform from saving an invalid new output value in the state. You can also use them to preserve a valid output value from the previous apply, if applicable." but it doesn't have an example :/

I can take that sentence out if you don't feel it's correct?

Author: Rita was specifically commenting about this last part of the sentence, saying it felt like an anti-pattern to promote:

"You can use preconditions on outputs to prevent Terraform from saving an invalid value in your state, **or to preserve a valid value from the previous apply.**"

This feels like an SME-y question, @BrianMMcClain or @robin-norwood do you have opinions here? I can chop off the last half of the sentence if we don't want to promote saving old state values.

Author: I realized that the original sentence doesn't actually cover what matters here, because the _point_ of an output is the data you are exposing from this module. So, what the user cares about is that you can use preconditions to establish your output's requirements before Terraform exposes those outputs. 

Are folks cool if I go with this instead?
```suggestion
An `output` block can also include a precondition to verify a module's output. Use preconditions on outputs to validate that your output values meets your requirements before Terraform exposes them.
```

Use the `check` block to validate your infrastructure outside of the typical resource lifecycle.

The `check` block executes as the last step of plan or apply operation, after Terraform has planned or provisioned your infrastructure. When a `check` block's assertion fails, Terraform reports a warning and continues executing the current operation.

Use `check` blocks when you want to:
- Validate resources, data sources, variables, or outputs in your configuration
- Validate the behavior of your infrastructure as a whole
- Verify infrastructure configuration without blocking operations
- Perform [continuous validation in HCP Terraform](#continuous-validation-in-hcp-terraform)

The following example uses a `check` block to verify the Terraform website is healthy.

```hcl
check "health_check" {
  data "http" "terraform_io" {
    url = "https://www.terraform.io"
  }

  assert {
    condition = data.http.terraform_io.status_code == 200
    error_message = "${data.http.terraform_io.url} returned an unhealthy status code"
  }
}
```

> If the website's endpoint returns a `200` status code, then the website is healthy and the check passes. Unlike other ways of validating your configuration, `check` blocks do not block operations. If the assertion evaluates to `false`, Terraform throws a warning that includes the result of the `error_message` expression and continues the operation.

Author: I think this is **super** important advantage of check blocks v.s. other ways of validating config, so I'm going to leave it in.

For more details, refer to the [`check` configuration reference](/terraform/language/block/check).

## Continuous validation in HCP Terraform

If you enable health checks on a workspace, HCP Terraform continuously validates any `check` blocks, preconditions, and postconditions in your workspace's configuration to regularly verify your infrastructure. For example, you can use a `check` block to continuously monitor the validity of an API gateway certificate. Continuous validation alerts you when the condition fails, so you can update the certificate and avoid errors the next time you want to update your infrastructure.

Refer to [Continuous validation](/terraform/cloud-docs/workspaces/health#continuous-validation) for more details.

## Order of validation

Terraform validates different aspects of your configuration as early as it can. Generally, Terraform executes evaluations in the following order:

1. Terraform executes input variable validations immediately, before generating a plan.

> 1. Terraform executes preconditions after generating a plan, but before creating the resource, data source, or output.

Reviewer: ```suggestion
1. Terraform executes preconditions after generating a plan but before creating the resource, data source, or output.
```
```suggestion
1. Terraform executes preconditions after generating a plan, but before creating the resource, data source, or output.
```

1. Terraform executes postconditions after planning and applying changes.

> 1. Terraform executes checks at the end of plan and apply operations, and whenever health assessments run on a workspace in HCP Terraform.

Reviewer: ```suggestion
1. Terraform executes checks at the end of plan and apply operations and every time health assessments run on a workspace in HCP Terraform.
```

![Order of operations for when validations run](/img/docs/validation-order-of-operations-dark.jpg#light-theme-only)

![Order of operations for when validations run](/img/docs/validation-order-of-operations-dark.jpg#dark-theme-only)

> The precise order that Terraform executes `check` blocks, preconditions, and postconditions can depend on whether Terraform knows a condition's value before or after applying your configuration. If the relevant value is available before an apply operation, then Terraform performs the validation during the planning phase. For example, if Terraform knows a resource's image ID during planning, it can execute any validations that rely on that ID.

Reviewer: ```suggestion
The precise order that Terraform executes `check` blocks, preconditions, and postconditions can depend on whether Terraform has information about a condition's value before or after applying your configuration. If the relevant value is available before an apply operation, then Terraform performs the validation during the planning phase. For example, if Terraform logs a resource's image ID during planning, it can execute any validations that rely on that ID.
```
We need to figure out more precise language about what it means for TF to "know" something so that we can edit out personification from the docs. Is the information logged internally? Is it captured in state? Is it cached or written to a temporary file?

Author: Good point! I think it's different at different points, "logs" isn't the _quite_ right word because what matters is if TF has access to the value. Accepting with a slight tweak:

```suggestion
The precise order that Terraform executes `check` blocks, preconditions, and postconditions can depend on whether Terraform has information about a condition's value before or after applying your configuration. If the relevant value is available before an apply operation, then Terraform performs the validation during the planning phase. For example, if Terraform has access to a resource's image ID during planning, it can execute any validations that rely on that ID.
```

If Terraform only knows the value after applying, then Terraform delays checking that validation rule until the apply phase. For example, AWS assigns the root volume ID when it starts an EC2 instance, so Terraform cannot reference the root volume ID until the apply is complete.

During the apply phase, a failed precondition prevents Terraform from implementing planned actions for the associated resource, data source, or output. A failed postcondition halts processing and prevents further downstream actions that rely on the resource or data source, but does not undo any actions Terraform has already taken.

## Error Messages

<>{/* TODO: remove once we have an input variable validation page, because then we'll have error_message described on each reference page and don't need this  */}</>

> Input variable validations, preconditions, postconditions, and checks all must include the `error_message` argument. This contains the text that Terraform will include as part of error messages when it detects an unmet condition.

Reviewer: ```suggestion
Input variable validations, preconditions, postconditions, and checks all must include the `error_message` argument. This contains the text that Terraform includes as part of error messages when it detects an unmet condition.
```

```
Error: Resource postcondition failed

  with data.aws_ami.example,
  on ec2.tf line 19, in data "aws_ami" "example":
  72:       condition     = self.tags["Component"] == "nomad-server"
    |----------------
    | self.tags["Component"] is "consul-server"

The selected AMI must be tagged with the Component value "nomad-server".
```

> The `error_message` argument can be any expression that evaluates to a string.

Reviewer: ```suggestion
The `error_message` argument supports any expression that evaluates to a string.
```
Side note, I like this way of framing data types. I've struggled with how to describe supported data types when you would commonly set the argument to a function or expression that _resolves_ to the data type Terraform requires.

Author: This was from the original article!

> This includes literal strings, heredocs, and template expressions. You can use the [`format` function](/terraform/language/functions/format) to convert items of `null`, `list`, or `map` types into a formatted string. Multi-line
> error messages are supported, and lines with leading whitespace will not be
> word wrapped.

Reviewer: ```suggestion
This includes literal strings, heredocs, and template expressions. You can use the [`format` function](/terraform/language/functions/format) to convert items of `null`, `list`, or `map` types into a formatted string. Multi-line
error messages are supported, and Terraform does not wrap  lines with leading whitespace.
```

We recommend writing error messages as one or more full sentences in a

> style similar to Terraform's own error messages. Terraform will show the
> message alongside the name of the resource that detected the problem and any
> external values included in the condition expression.

Reviewer: ```suggestion
style similar to Terraform's own error messages. Terraform shows the
message alongside the name of the resource that detected the problem and any
external values included in the condition expression.
```
